cartridge/roles/expirationd.lua

```diff
@@ -176,6 +210,15 @@ local function apply_config(conf_new, opts)
         table.remove(started, i)
     end
 
+    if conf["cfg"] ~= nil then
+        local ok = pcall(get_task_config, conf["cfg"])
+        if not ok then
+            local cfg = get_cfg(conf["cfg"])
+            expirationd.cfg(cfg)
```

Comment:
I think it's a bit trickier than this.

After this patch, user can either change configuration with `expirationd.cfg()`, which changes `expirationd.cfg` table, or with clusterwide config, which changes clusterwide config and `expirationd.cfg` table. It may be confusing that the contents of the clusterwide config section are not guaranteed to be the same as expirationd config table. This state can easily be achieved with calling `expirationd.cfg()` after applying something with clusterwide config. Also it is possible that some value X would be set only with `expirationd.cfg()` and never set explicitly in clusterwide config (it is not possible for current expirationd cfg since for now there is only one value -- `metrics`, but it is possible for `crud.cfg` which will implement similar features [soon](https://github.com/tarantool/crud/issues/332)). How can we treat it?
- The simplest case is to ignore it. Yeah, such behavior may be dissatisfying, but in most of the real cases configuration is not changed frequently (especially with two different handles). Then we must at least add a documentation warning that clusterwide config value may not represent the current `expiration.cfg{}` state and you should check `expirationd.cfg` attributes to get real values.
- We may make the behavior to be symmetric: `expirationd.cfg{}` will update clusterwide config. As far as I understand, this could be achieved with [`patch_clusterwide`](https://www.tarantool.io/ru/doc/latest/book/cartridge/cartridge_api/modules/cartridge.twophase/#patch-clusterwide-patch). The conditions of "when we should really update it" may differ: for example, we may update it if expirationd role is enabled, or if the role is enabled and there is `expirationd` section, or if the role is enabled and there is `expirationd.cfg` subsection. This one is also tricky since it is possible that `expirationd.cfg` was called before the role was enabled (but it's more about the case when we have several fields in configuration). Also I'm not sure that patching clusterwide config from the instance is a really good design of a cartridge application.
- We may simply throw error on `expirationd.cfg` call if there is a clusterwide configuration subsection `expirationd.cfg`, asking a user to continue to use clusterwide config. But it could lead to major issues on migration from `expirationd.cfg` to clusterwide config.

I see this task to be a bit more than "allow an expirationd user to set cfg with clusterwide config" -- it is likely to become an exemplar approach on how to merge `module.cfg` with clusterwide config (at least for similar task in `crud.cfg`). Maybe we should discuss this with TeamU to hear their opinion.

Reply:
As far as I understand, there are no similar examples yet -- `metrics` has `metrics.set_export`, but this handle is expected to be called at most once and has lower priority than clusterwide config (current `module.cfg` has higher priority and to change this is to introduce a breaking change).

Reply:
> After this patch, user can either change configuration with `expirationd.cfg()`, which changes `expirationd.cfg` table, or with clusterwide config, which changes clusterwide config and `expirationd.cfg` table.

My logic here is very simple:
1. clusterwide config is a global configuration to set up for all nodes by default.
2. expirationd.cfg() is a for a user/node-specific/manual configuration.

So there is nothing confusing in the fact that the values from clusterwide config and expirationd config are different.

> We may make the behavior to be symmetric: expirationd.cfg{} will update clusterwide config. As far as I understand, this could be achieved with [patch_clusterwide](https://www.tarantool.io/ru/doc/latest/book/cartridge/cartridge_api/modules/cartridge.twophase/#patch-clusterwide-patch). The conditions of "when we should really update it" may differ: for example, we may update it if expirationd role is enabled, or if the role is enabled and there is expirationd section, or if the role is enabled and there is expirationd.cfg subsection.

Here, too, everything is simple:

We may update the config only if the update is called from the role: cartridge.service_get('expirationd').cfg({}). But this is a bad idea - the configuration needs to be updated from the top-level, not from the low-level.

> We may simply throw error on expirationd.cfg call if there is a clusterwide configuration subsection expirationd.cfg, asking a user to continue to use clusterwide config. But it could lead to major issues on migration from expirationd.cfg to clusterwide config.

Ideally, the application should not know anything about the cartridge. We can only do this from the role (and I don't see the point in this restriction).

Reply:
>  but it is possible for `crud.cfg` which will implement similar features [soon](https://github.com/tarantool/crud/issues/332)). 

For which values will this be relevant and why?

Reply:
As a result:

1. A user sets the global configuration that the cartridge accepts for the module per each instance.
2. After that a user can set a local module configuration per a specific instance.

I don't see any logical problems in this logic. Yeah, the user will need to manually re-apply the local configuration of the module after cargirdge's apply_config(), but this is quite expected.

Reply:
Most people upload configuration to set up something, but it is also common to download configuration/inspect it in UI to see how everything is configured now. Current solution does not allow to do the same for expirationd since clusterwide configuration info may not be related to real cfg at all.

Reply:
> > but it is possible for `crud.cfg` which will implement similar features [soon](https://github.com/tarantool/crud/issues/332)).
> 
> For which values will this be relevant and why?

One may set up `metrics` in `crud.cfg`, but use clusterwide configuration to set `metrics_driver`.

Reply:
But your point about configuring instances differently is a valid one. Thus, it seems that we will stick to the first approach. I think we still need to add documentation warnings about configuration priority and overriding.

Reply:
> I think we still need to add documentation warnings about configuration priority and overriding.

Added:

![123](https://user-images.githubusercontent.com/29621138/211837403-6ce99dc4-28df-4b5a-a53c-0ea245d9fc16.png)